flash/core/finetuning.py

```diff
@@ -0,0 +1,142 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import inspect
+from typing import List, Union
+
+import pytorch_lightning as pl
+from pytorch_lightning.callbacks import BaseFinetuning
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from torch import nn
+from torch.optim import Optimizer
+
+_EXCLUDE_PARAMTERS = ("self", "args", "kwargs")
+
+
+class NoFreeze(BaseFinetuning):
+    pass
+
+
+class FlashBaseBaseFinetuning(BaseFinetuning):
+
+    def __init__(self, attr_names: Union[str, List[str]] = "backbone", train_bn: bool = True):
+        self.attr_names = [attr_names] if isinstance(attr_names, str) else attr_names
+        self.train_bn = train_bn
+
+    def freeze_before_training(self, pl_module: pl.LightningModule) -> None:
+        self.freeze_using_attr_names(pl_module, self.attr_names, train_bn=self.train_bn)
+
+    @staticmethod
+    def freeze_using_attr_names(pl_module, attr_names: List[str], train_bn: bool = True):
+        for attr_name in attr_names:
+            attr = getattr(pl_module, attr_name, None)
+            if attr is None or not isinstance(attr, nn.Module):
+                MisconfigurationException("To use Freeze your model must have a {attr} attribute")
+            BaseFinetuning.freeze(module=attr, train_bn=train_bn)
```

Comment:
I don't think you want this to be a static call.

Because if somebody overrides freeze, this will ignore it.

Reply:
Yes, but this code currently lives in Lightning. However, if a user is overriding freeze, it should know what he is doing at this point.

Reply:
I'm not getting you. A flash user could do

```python
class MyFreezeUnfreeze(FreezeUnfreeze):
    def freeze(self, *args, **kwargs):
        do_something()
        super().freeze(*args, **kwargs)
```

And then `do_something` wouldn't be called because you are calling `BaseFinetuning.freeze`

So he would also have to override `freeze_using_attr_names` to get his code to work

Reply:
Fine.